### PR TITLE
Add ToC cookbook recipe

### PR DESCRIPTION
**Pull request by texnixe**

### CODE REVIEW

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+----
+
+Text:
+
+A Table of Contents (ToC) gives readers a quick overview of the contents of a page and allows accessing specific parts quickly.
+
+## Prerequisites
+Being familiar with (link: docs/guide/plugins/plugin-basics text: registering extensions in Kirby) and a basic understanding of PHP will be useful.  
+
+To follow along, (link: docs/guide/quickstart#download-and-installation text: install a Kirby Starterkit), create a `/toc` folder with an `index.php` inside `/site/plugins/` and add some headlines in a textarea field, for example in one of the subpages of the `/notes` page.
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]:
```suggestion
To follow along, (link: docs/guide/quickstart#download-and-installation text: install a Kirby Starterkit), create a `/site/plugins/toc` folder with an `index.php` inside and add some headlines in a textarea field, for example in one of the subpages of the `/content/notes` page.
```

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+<info>
+Note that the techniques described here only work for fields that are rendered using the `kirbyText()` method, usually a `textarea` field (or plugin fields derived from this), not with the (link: https://github.com/getkirby/editor/ text: Editor).
+</info>
+
+## Plugin parts
+
+To create a ToC, we need two things:
+
+- Anchors in the text to which the ToC can link. While those could be added manually, it is much easier to auto-generate them from headlines. Anchors are useful even without a ToC, because you can link to these anchors from elsewhere.
+- A navigation list of links to these anchors usually at the top of the page or the sidebar, the actual ToC
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]:
```suggestion
- A navigation list of links to these anchors usually at the top of the page or the sidebar, the actual ToC.
```

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+
+A Table of Contents (ToC) gives readers a quick overview of the contents of a page and allows accessing specific parts quickly.
+
+## Prerequisites
+Being familiar with (link: docs/guide/plugins/plugin-basics text: registering extensions in Kirby) and a basic understanding of PHP will be useful.  
+
+To follow along, (link: docs/guide/quickstart#download-and-installation text: install a Kirby Starterkit), create a `/toc` folder with an `index.php` inside `/site/plugins/` and add some headlines in a textarea field, for example in one of the subpages of the `/notes` page.
+
+<info>
+Note that the techniques described here only work for fields that are rendered using the `kirbyText()` method, usually a `textarea` field (or plugin fields derived from this), not with the (link: https://github.com/getkirby/editor/ text: Editor).
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]:
```suggestion
Note that the techniques described here only work for fields that are rendered using the `kirbytext()` method, usually a `textarea` field (or plugin fields derived from this), not with the (link: https://github.com/getkirby/editor/ text: Editor).
```

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+We will look into both options in this recipe.
+
+<info>
+To keep it simple, we will include only one headline level in our ToC. We will cover multiple levels in a future recipe.
+</info>
+
+## Using a `toc` placeholder
+
+For the `toc` placeholder, we use another `kirbyText:after` hook, and a snippet that will contain the HTML for the ToC. We will reuse this snippet for the (link: #alternative-a-field-method-to-generate-the-toc text: alternative field method approach) below.
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]:
```suggestion
For the `toc` placeholder, we use another `kirbytext:after` hook, and a snippet that will contain the HTML for the ToC. We will reuse this snippet for the (link: #alternative-a-field-method-to-generate-the-toc text: alternative field method approach) below.
```

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+  'hooks' => [
+  
+    'kirbytext:after' => [
+
+      // (…) add the callback function from above here (without the wrapper)
+        
+      function($text) {
+
+        // the pattern allows passing an optional headline level `(toc: h3)`
+        $pattern  = '!\(toc:?\s?(h[1-6])?\)!';
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]:
```suggestion
        $pattern  = '!\(toc(?::\s?(h[1-6]))?\)!';
```

(the colon is only optional when no parameter is passed – needs to be tested though if it still fully works)

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        $text = preg_replace_callback($pattern, function($match) use($text) {
+          
+          // get the headline level from the match
+          $headline = $match[1] ?? 'h2';
+
+          // find all headline matches
+          preg_match_all('!<' . $headline . '.*?>(.*?)</' . $headline . '>!s', $text, $matches);
+
+          // create a new collection for the headlines…
+          $headlines = new Collection;
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]:
```suggestion
          $headlines = new Collection();
```

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+  </nav>
+<?php endif ?>
+```
+
+### Result
+If you now add a `(toc)` placeholder to a page that contains headlines of the given level, you should see a ToC appear where you added the placeholder.
+
+## Alternative: A field method to generate the anchors
+
+This is an alternative to using the `kirbyText` hook described above. Do not use the hook and this method in conjuntion.
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]:
```suggestion
This is an alternative to using the `kirbytext` hook described above. Do not use the hook and this method in conjunction.
```

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+    </ol>
+
+  </nav>
+<?php endif ?>
+```
+
+### Result
+If you now add a `(toc)` placeholder to a page that contains headlines of the given level, you should see a ToC appear where you added the placeholder.
+
+## Alternative: A field method to generate the anchors
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]: I’d move this section up to the anchors. I first thought this was a ToC alternative, so I was confused reading this below the ToC section.

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+```php "/site/plugins/toc/index.php"
+<?php
+
+Kirby::plugin('k-cookbook/toc', [
+  'fieldMethods' => [
+    'headlines' => function($field, $headline = 'h2') {
+
+        preg_match_all('!<' . $headline . '.*?>(.*?)</' . $headline . '>!s', $field->kt()->value(), $matches);
+
+        $headlines = new Collection;
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]:
```suggestion
        $headlines = new Collection();
```

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+    }
+  ]
+]);
+```
+The field method returns a collection of headline objects.
+
+In your templates, you can now use the `headlines()` field method in conjunction with the snippet we registered above.
+
+```php
+<?php snippet('toc', ['headlines' => $page->text()->kt()->headlines('h2')]) ?>
```
Comment (lukasbestle) [inline content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]: Doesn’t this need to be this as the method also calls `kt()`?

```suggestion
<?php snippet('toc', ['headlines' => $page->text()->headlines('h2')]) ?>
```

**Review by lukasbestle — COMMENTED** [review]
I really like the detailed explanations. 👍

File: content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt (hunk 10 above)
Comment (texnixe) [reply content/1_docs/2_cookbook/7_extensions/0_table-of-contents/recipe.txt]: Of course!

**Pull request merged**